rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,161 @@
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+- OAuth client ID and secret
+
+This object has very strict access control requirements that will be explained in detail
+in the **Access Control** section.
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: plugin-name # name must match the name of the plugin.
+spec:
+  last_rotated: "2023-05-09T16:50:52.497874Z"
+  credentials:
+    # Only one of these credential types must be defined at a time. All values in this
+    # are string literals and not file locations.
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+    oauth_client_secret:
+      client_id: example-client-id
+      client_secret: example-client-secret
+```
+
+If a plugin requires static credentials, there is a 1-to-1 mapping between `Plugin`
+objects and `PluginStaticCredentials` objects. If a `Plugin` is using oauth, however,
+it will not need a `PluginStaticCredentials` and none will be created. The object will
+inherit the same name as its associated plugin.
+
+### Changes to the `Plugin` object
+
+The `Plugin` object's `Credentials` field will have a new type of credential called
+`PluginNeedsStaticCredentials` with a single boolean `Enabled` flag. If this flag is set,
+this will indicate to Teleport that the plugin must have a corresponding
+`PluginNeedsStaticCredentials` object in order to start.
+
+### Access control
+
+`PluginStaticCredentials` can be written to and deleted by any user with RBAC permissions
+to do so. For these operations, the object is straightforward and reflects the typical
+way that Teleport resources are handled.
+
+However, the object will be only readable by the `RoleAdmin` role (with one exception).
+This is because plugins run on the auth server, so it's the only role that needs to be
+able to actually read the static credentials.
+
+The one exception here is Teleport Assist, which runs in the proxy. The proxy will be
+able to read plugin credentials with the name of `openai-default`, which is the name of
+the Assist plugin.
+
+### Lifecycle of the `PluginStaticCredentials` object
+
+#### Creation
+
+A `PluginStaticCredentials` object can be created after a `Plugin` object has been
+created. If a `Plugin` is marked as needing static credentials as described above, the
+`Plugin` won't start until these have been provided.
+
+When creating a plugin, Teleport will first perform a lookup to ensure that there is a
+plugin with the same name as the new `PluginStaticCredentials` object being created
+and that the plugin has `PluginNeedsStaticCredentials` with the `Enabled` flag set to
+true. If this plugin doesn't exist, the creation of the `PluginStaticCredentials` object
+will fail.
+
+#### Deletion
+
+On deletion, the associated `Plugin` will be stopped. Additionally, `PluginStaticCredentials`
```

Comment:
The cascading delete might be a bit surprising if/when we get a workflow for manually onboarding plugins via tctl or similar. Take the scenario:
1. Create Plugin
2. Create PluginStaticCreds
3. Delete Plugin with intent to re-create
4. Create Plugin. Plugin does not start because PluginStaticCreds have been silently deleted.

I guess the main goal here is to not accidentally hold on to any credentials longer than they are needed?

Reply:
Yes, that's correct. I've adjusted this lifecycle slightly so that this makes more sense (IMO).